### PR TITLE
Multiple problems support

### PR DESCRIPTION
**Pull request by somnath77**
1. Codes autosave apis
2. changes in submissions and problems apis to support multiple problems

### CODE REVIEW

File: app/controllers/api/v1/candidates_controller.rb [hunk 1]
```diff
@@ -91,14 +91,22 @@ def check_emails_present
...
 
         start_time = DateTime.now.localtime
-        end_time = start_time + drive_problem.problem.time_in_minutes.minutes
+        end_time = start_time + total_time(drive_problems).minutes
         drives_candidate.update(start_time: start_time, end_time: end_time)
       end
+
+      def total_time(drive_problems)
+        total_time = 0
+        drive_problems.each do |drive_problem|
```
Comment (anilmaurya) [inline app/controllers/api/v1/candidates_controller.rb:105]:
```suggestion
        Problem.where(id: drive_problems.collect(&:problem_id)).collect(&:time_in_minutes).sum
```

File: app/controllers/api/v1/codes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+module Api
+  module V1
+    class CodesController < ApiController
+      before_action :find_drives_candidate, :find_problem
+      before_action :find_code, only: %i[show]
+
+      def create
+        code = Code.find_or_initialize_by(drives_candidate_id: @drives_candidate.id, problem_id: @problem.id)
+        code.save
```
Comment (anilmaurya) [inline app/controllers/api/v1/codes_controller.rb:11]: save not required

File: app/controllers/api/v1/problems_controller.rb [hunk 3]
```diff
@@ -3,18 +3,22 @@
...
+      before_action :load_drive
+
       def index
-        if (problem = @drive_problem.problem)
-          render_success(data: problem, message: I18n.t('success.message'))
-        else
-          render_error(message: I18n.t('not_found.message'), status: 404)
+        drives_problems = DrivesProblem.where(drive_id: @drive.id)
+        problems = []
+        drives_problems.each do |drives_problem|
```
Comment (anilmaurya) [inline app/controllers/api/v1/problems_controller.rb:11]: use collect instead of each

File: app/jobs/submission_job.rb [hunk 4]
```diff
@@ -33,7 +33,13 @@ def run_testcase(testcase)
 
   def calculate_result
     drives_candidate = @submission.drives_candidate
-    highest_submission = drives_candidate.submissions.order('total_marks desc').first
-    drives_candidate.update(score: highest_submission.total_marks)
+    problems = drives_candidate.drive.problems
+    score = 0
+    problems.each do |problem|
```
Comment (anilmaurya) [inline app/jobs/submission_job.rb:38]: Use collect and sum

File: db/migrate/20210713060739_remove_drives_problem_from_codes.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class RemoveDrivesProblemFromCodes < ActiveRecord::Migration[6.1]
+  def change
+    remove_reference :codes, :drives_problem, null: false, foreign_key: true
```
Comment (anilmaurya) [inline db/migrate/20210713060739_remove_drives_problem_from_codes.rb:5]: Rollback to previous migration and make all changes related to codes in one migration

File: spec/acceptance/api/v1/code_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    parameter :token, 'DrivesCandidate token'
+    parameter :problem_id, 'Problem id'
+    parameter :answer, 'Code answer'
+    parameter :language_id, 'Code lang_code'
+    context 'with valid params' do
+      let!(:token) { drives_candidate.token }
+      let!(:problem_id) { problem.id }
+      let!(:answer) { Faker::Lorem.paragraph }
+      let!(:language_id) { Faker::Number.digit }
+      example 'create code for a problem and store to database' do
```
Comment (anilmaurya) [inline spec/acceptance/api/v1/code_controller_spec.rb:58]: Check if Code count is 0

File: spec/acceptance/api/v1/code_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    context 'with invalid params' do
+      let!(:token) { Faker::Internet.uuid }
+      let!(:problem_id) { problem.id }
+      let!(:answer) { Faker::Lorem.paragraph }
+      let!(:language_id) { Faker::Number.digit }
+      example ' returns not found error message as token is fake' do
+        do_request
+        response = JSON.parse(response_body)
+        expect(response['message']).to eq(I18n.t('not_found.message'))
+        expect(status).to eq(200)
```
Comment (anilmaurya) [inline spec/acceptance/api/v1/code_controller_spec.rb:75]: check code count is 0

Comment (somnath77) [reply spec/acceptance/api/v1/code_controller_spec.rb:75]: done

File: spec/acceptance/api/v1/code_controller_spec.rb (hunk 6 above)
Comment (somnath77) [reply spec/acceptance/api/v1/code_controller_spec.rb:58]: done

File: db/migrate/20210713060739_remove_drives_problem_from_codes.rb (hunk 5 above)
Comment (somnath77) [reply db/migrate/20210713060739_remove_drives_problem_from_codes.rb:5]: done

File: app/controllers/api/v1/problems_controller.rb (hunk 3 above)
Comment (somnath77) [reply app/controllers/api/v1/problems_controller.rb:11]: done

File: app/controllers/api/v1/codes_controller.rb (hunk 2 above)
Comment (somnath77) [reply app/controllers/api/v1/codes_controller.rb:11]: done

File: app/controllers/api/v1/candidates_controller.rb (hunk 1 above)
Comment (somnath77) [reply app/controllers/api/v1/candidates_controller.rb:105]: done

File: app/jobs/submission_job.rb (hunk 4 above)
Comment (somnath77) [reply app/jobs/submission_job.rb:38]: Here using collect is not straight forward as  each problem has multiple submissions, and we need one submissions with maximum marks

Comment (anilmaurya) [reply app/jobs/submission_job.rb:38]:
```suggestion
    score = problems.collect do |problem|
      submissions = Submission.where(drives_candidate_id: drives_candidate.id, problem_id: problem.id)
      highest_submission = submissions.order('total_marks desc').first
      highest_submission.total_marks
    end.sum
```
won't this work ? @somnath77

Comment (somnath77) [reply app/jobs/submission_job.rb:38]: It worked, thanks

**Pull request merged**